File: javascript/squares-numbers-sorted-array.js
/**
 * Solution works with O(N) two pointers
 * @url https://leetcode.com/explore/learn/card/fun-with-arrays/521/introduction/3240/
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function (nums) {
  const result = new Array(nums.length);
  let left = 0;
  let right = nums.length - 1;
  let size = nums.length - 1;

  for (let i = size; i >= 0; i--) {
    if (Math.abs(nums[left]) >= Math.abs(nums[right])) {
      result[i] = nums[left] ** 2;
      left++;
    } else {
      result[i] = nums[right] ** 2;
      right--;
    }
  }

  return result;
};

File: python/squares-numbers-sorted-array.py
from typing import List


def sortedSquares(nums: List[int]) -> List[int]:
  result = [None] * len(nums)
  left = 0
  right = len(nums) - 1
  
  # i >= 0
  for i in range(len(nums) - 1, -1, -1):
    print(i, nums[right], nums[left])

    if abs(nums[left]) >= abs(nums[right]):
      result[i] = nums[left] ** 2
      left += 1
    else:
      result[i] = nums[right] ** 2
      right -= 1

  return result

print(sortedSquares([-4,-1,0,3,10]))
        